lib/evaluate/integer.h

```diff
@@ -0,0 +1,853 @@
+  static_assert(parts >= 1);
+  static constexpr int extraTopPartBits{
+      extraPartBits + (parts * partBits) - bits};
+  static constexpr int topPartBits{maxPartBits - extraTopPartBits};
+  static_assert(topPartBits > 0 && topPartBits <= partBits);
+  static_assert((parts - 1) * partBits + topPartBits == bits);
+  static constexpr Part partMask{static_cast<Part>(~0) >> extraPartBits};
+  static constexpr Part topPartMask{static_cast<Part>(~0) >> extraTopPartBits};
+
+public:
+  struct ValueWithOverflow {
+    Integer value;
+    bool overflow;
+  };
+
+  struct ValueWithCarry {
+    Integer value;
+    bool carry;
+  };
+
+  struct Product {
+    Integer upper, lower;
+  };
+
+  struct QuotientWithRemainder {
+    Integer quotient, remainder;
+    bool divisionByZero, overflow;
+  };
+
+  // Constructors and value-generating static functions
+  constexpr Integer() { Clear(); }  // default constructor: zero
+  constexpr Integer(const Integer &) = default;
+  constexpr Integer(std::uint64_t n) {
+    for (int j{0}; j + 1 < parts; ++j) {
+      SetLEPart(j, n);
+      if constexpr (partBits < 64) {
+        n >>= partBits;
+      } else {
+        n = 0;
+      }
+    }
+    SetLEPart(parts - 1, n);
+  }
+  constexpr Integer(std::int64_t n) {
+    std::int64_t signExtension{-(n < 0)};
+    signExtension <<= partBits;
+    for (int j{0}; j + 1 < parts; ++j) {
+      SetLEPart(j, n);
+      if constexpr (partBits < 64) {
+        n = (n >> partBits) | signExtension;
+      } else {
+        n = signExtension;
+      }
+    }
+    SetLEPart(parts - 1, n);
+  }
+
+  constexpr Integer &operator=(const Integer &) = default;
+
+  // Left-justified mask (e.g., MASKL(1) has only its sign bit set)
+  static constexpr Integer MASKL(int places) {
+    if (places <= 0) {
+      return {};
+    } else if (places >= bits) {
+      return MASKR(bits);
+    } else {
+      return MASKR(bits - places).NOT();
+    }
+  }
+
+  // Right-justified mask (e.g., MASKR(1) == 1, MASKR(2) == 3, &c.)
+  static constexpr Integer MASKR(int places) {
+    Integer result{nullptr};
+    int j{0};
+    for (; j + 1 < parts && places >= partBits; ++j, places -= partBits) {
+      result.LEPart(j) = partMask;
+    }
+    if (places > 0) {
+      if (j + 1 < parts) {
+        result.LEPart(j++) = partMask >> (partBits - places);
+      } else if (j + 1 == parts) {
+        if (places >= topPartBits) {
+          result.LEPart(j++) = topPartMask;
+        } else {
+          result.LEPart(j++) = topPartMask >> (topPartBits - places);
+        }
+      }
+    }
+    for (; j < parts; ++j) {
+      result.LEPart(j) = 0;
+    }
+    return result;
+  }
+
+  static constexpr ValueWithOverflow ReadUnsigned(
+      const char *&pp, std::uint64_t base = 10) {
+    Integer result;
+    bool overflow{false};
+    const char *p{pp};
+    while (*p == ' ' || *p == '\t') {
+      ++p;
+    }
+    if (*p == '+') {
+      ++p;
+    }
+    Integer radix{base};
```

Comment:
Do we need a warning comment for people in the future with non-linear character sets?  An assertion that base is less than 37 or something?

Reply:
disclaimer added